### ISSUE TITLE
loss calculate

### ISSUE THREAD

**Issue by nikefd** [facebookresearch/fairseq-lua#70]
Sorry to bother you again. I am also confused with the loss function.
Can you make an brief explanation about how it calculate?

I see that you set the criterion in the file `train.lua` and pass the parameter to ResumableDPOptimEngine.
`local criterion = nn.CrossEntropyCriterion(critweights, false):cuda()` and I am lost in the file==.

After some steps,  I found that 
```
runTest(state, set, meter)
local loss = meter:value() / math.log(2)
```

I guess that we calculate loss use the cross-entropy cost function of every batch. And add them together. 
And do something others.

**Comment by jgehring** [comment]
You're on the right track. The criterion computes the cross-entropy loss on forward(), and the loss of each worker is summed up in the engine when training (see https://github.com/facebookresearch/fairseq/blob/master/fairseq/torchnet/ResumableDPOptimEngine.lua#L375-L378 for the relevant lines). Please note that we compute the **sum** of the loss over all tokens, and thanks to the criterion weights we ignore padding tokens in the computation. The `updateMeters` hook is called after each sample and updates the meters with the latest loss value (https://github.com/facebookresearch/fairseq/blob/master/fairseq/torchnet/hooks.lua#L143). We then log the average loss over all non-padding tokens in the current epoch in https://github.com/facebookresearch/fairseq/blob/master/train.lua#L356.

**Comment by nikefd** [comment]
Thank you very much. Got them.
